go/vt/vtctl/reparentutil/replication.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reparentutil
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"vitess.io/vitess/go/event"
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/vt/concurrency"
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topotools/events"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vttablet/tmclient"
+
+	replicationdatapb "vitess.io/vitess/go/vt/proto/replicationdata"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+// FindValidEmergencyReparentCandidates will find candidates for an emergency
+// reparent, and, if successful, return a mapping of those tablet aliases (as
+// raw strings) to their replication positions for later comparison.
+func FindValidEmergencyReparentCandidates(
+	statusMap map[string]*replicationdatapb.StopReplicationStatus,
+	primaryStatusMap map[string]*replicationdatapb.MasterStatus,
+) (map[string]mysql.Position, error) {
+	replicationStatusMap := make(map[string]*mysql.ReplicationStatus, len(statusMap))
+	positionMap := make(map[string]mysql.Position)
+
+	// Build out replication status list from proto types.
+	for alias, statuspb := range statusMap {
+		status := mysql.ProtoToReplicationStatus(statuspb.After)
+		replicationStatusMap[alias] = &status
+	}
+
+	// Determine if we're GTID-based. If we are, we'll need to look for errant
+	// GTIDs below.
+	var (
+		isGTIDBased                bool
+		isNonGTIDBased             bool
+		emptyRelayPosErrorRecorder concurrency.FirstErrorRecorder
+	)
+
+	for alias, status := range replicationStatusMap {
+		if _, ok := status.RelayLogPosition.GTIDSet.(mysql.Mysql56GTIDSet); ok {
+			isGTIDBased = true
+		} else {
+			isNonGTIDBased = true
+		}
+
+		if status.RelayLogPosition.IsZero() {
+			// Potentially bail. If any other tablet hits the non-default
```

Comment:
It took me a minute to understand this, it isn't clear what non-default means. Can you rephrase? Something like this (or your own phrasing is fine too)
```
// Record an error that we will only return if we have a mix of GTID and non-GTID tablets
```

Reply:
I still don't think I understood it correctly, so it's up to you to write the correct comment :)

Reply:
Yep, sorry, this was written at the time when I was using a switch case with a default here, and I forgot to go back and update the comment. Will fix.